Add docs on dumps (#21704)
dotnet/docs: docs/core/diagnostics/dumps.md

---
title: Dumps - .NET
description: An introduction to dumps in .NET.
ms.date: 10/12/2020
---

# Dumps

> A dump is a file that contains snapshot of the process at the time it was created and can be useful for examining the state of your application. Dumps can be used for post-mortem debugging of your .NET application when it is difficult to attach a debugger to it such as production or CI environments. Using dumps allows you to capture the state of the problematic process and examine it without having to stop the application.

Reviewer: ```suggestion
A dump is a file that contains snapshot of the process at the time it was created and can be useful for examining the state of your application. Dumps can be used to debug your .NET application when it is difficult to attach a debugger to it such as production or CI environments. Using dumps allows you to capture the state of the problematic process and examine it without having to stop the application.
```

Dumps are more than just post mortem. It can be hangs and other stuff like generic probing.

> A dump is a file that contains snapshot of the process at the time they were created and can be useful for examining the state of your application for debugging. Dumps can be used for post-mortem debugging of your .NET application when it is difficult to attach a debugger to it (for example, when the app is running in production). Using dumps allows you to capture the state of the problematic process and examine it without having to stop it from running while being analyzed.

Reviewer: ```suggestion
A dump is a file that contains snapshot of the process at the time it was created and can be useful for examining the state of your application. Dumps can be used for post-mortem debugging of your .NET application when it is difficult to attach a debugger to it such as production or CI environments. Using dumps allows you to capture the state of the problematic process and examine it without having to stop the application.
```

Reviewer 2: ...contains a snapshot...

Reviewer: Caught it in the first. Slipped in the second. Thanks @gewarren

> ## Collect dumps

Reviewer: The TOC generated at the beginning of will use this title. Is something like "Collecting Dumps" or "How to Collect Dumps" more readable? I'm asking mostly from what's the usage in docs vs. what's more readable.

Reviewer 2: This is the correct format. We are not supposed to use -ing words (gerunds).

Dumps can be collected in a variety of ways depending on which platform you are running your app on.

> [!NOTE]

> > Collecting a dump inside a container requires PTRACE capability which can be added via `--cap-add=SYS_PTRACE` or `--privileged`.

Reviewer: Another note would be: know that these contain all the information in the process - handle with care as it could contain sensitive information.

### Using dotnet-dump

`dotnet-dump` is a CLI tool that allows you to collect and analyze dumps. For more information on how to use it to collect dumps with `dotnet-dump`, refer to [its documentation](dotnet-dump.md)

> ### Using environment variables

Reviewer: You can't really control when the environment variables capture dumps. It's only the crash case. The previous title should be "Collecting dumps at user's request". And point them to dotnet-dump, or the client library. For this one it should be Collecting dumps when the application crashes.

For some situations, it may be more desirable to use environment variables to collect dumps. For example, capturing a dump when an exception is thrown helps you identify issue by examining the state of the app when it crashed.

The table below shows the environment variables you can configure for collecting dumps.

|Environment variable|Description|Default value|
|-------|---------|---|
|`COMPlus_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|
|`COMPlus_DbgMiniDumpType`|Type of the dump to be collected. (See [Types of dumps](#types-of-dumps-in-.net) to learn more)|2 (`MiniDumpWithPrivateReadWriteMemory`)|
|`COMPlus_DbgMiniDumpName`|Path to a file to write the dump to.|`/tmp/coredump.<pid>`|
|`COMPlus_CreateDumpDiagnostics`|If set to 1, enable diagnostic logging of dump process.|0|

> ### Types of dumps in .NET

Reviewer: Conceptually this is a subtitle of the last bullet. It could have an extra level of the header, but i'd say it's enough to say "where COMPlus_DbgMiniDumpType is one of the following"

Author: I wanted to write it as a separate bullet because they can be used with `dotnet-dump` as well. Does that sound fine to you?

Reviewer: They don't map one to one, and the dotnet-dump article has it, with it's own blurb. If we are going to point them there, might as well not randomize them in this article anymore.

You can collect different types of the dump depending on the purpose. This can be configured with the `COMPlus_DbgMiniDumpType` when using environment variable, or the `--type` flag when you are using `dotnet-dump`. Refer to the table below on what types of dumps you can collect in .NET.

|Value|Name|Description|
|-----|----|-----------|
|1|`MiniDumpNormal`|Include just the information necessary to capture stack traces for all existing threads in a process. Limited GC heap memory and information.|
|2|`MiniDumpWithPrivateReadWriteMemory`|Includes the GC heaps and information necessary to capture stack traces for all existing threads in a process.|
|3|`MiniDumpFilterTriage`|Include just the information necessary to capture stack traces for all existing threads in a process. Limited GC heap memory and information.|
|4|`MiniDumpWithFullMemory`|Include all accessible memory in the process. The raw memory data is included at the end, so that the initial structures can be mapped directly without the raw memory information. This option can result in a very large file.|

## Analyze dumps

Dumps can be analyzed using [`dotnet-dump`](dotnet-dump.md).

> ## See More

Reviewer 2: ```suggestion
## See also
```
